anthony/unrealized-potential-of-string-diagrams.md

```diff
@@ -0,0 +1,128 @@
+---
+title: The Unrealized Potential of String Diagrams for ZKVMs
+publish_date: ???
+category: compilers
+image: media/string_puppet.webp
+excerpt: This post will describe string diagrams and explain why they are a well suited for designing ZKVMs.
+---
+
+## Trees Considered Harmful
+
+By the end of this post, I want to convince the reader that there are better forms of syntax than abstract syntax trees; namely, graphical ones. I'll also touch on how this may influence better ZKVM designs.
+
+First, *what's so bad about trees?* 
+
+Historically, most languages in mathematics have been constructed using abstract syntax trees, or ASTs. The very definition of a grammar in most computer science contexts assumes that any term of any language must be structured like a tree. Most of the time, this seems fine, but a century of hacks have had to be invented to compensate for the deficiencies of trees.
+
+I think the easiest place to see this is in the combinator calculus. Variables remain a persistant issue for theory, and much ink has been spilt trying to treat variables, or really, variable names, coherently. Issues related to scoping and shadowing formally complicate the syntactical aspects of mathematics which most simply take for granted. Among the earliest methods for managing them is to simply eliminate them. From this, we get the original combinator calculus invented by Skonfinkle. Conceptually similar systems were invented over the last century for similar reasons, such as [predicate functor logic](https://en.wikipedia.org/wiki/Predicate_functor_logic). These allow one to reason through symbol manipulation without needing named variables, but one must spend an inordinate effort dealing with a convoluted origami just to get separate ends of trees to communicate.
```

Comment:
several things

- only halfway through this paragraph I got it:  the idea of having no variables at all

- I would write Schönfinkel (and btw. Germans say Schönfinkeln for Currying) https://en.wikipedia.org/wiki/Moses_Sch%C3%B6nfinkel 

- can we have an example of some `let .. in ...` term and how the tree unfolds it or another concrete example, so that the inadequacy jumps to the eye ? 

- this reminds me of  Yiannis Moschovakis, using letrec as an extension of Montagovian semantics of natural language (just saying) https://www.math.ucla.edu/~ynm/lectures/es10.pdf